org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/internal/SSLNetworkModule.java

```diff
@@ -134,17 +136,25 @@ public void start() throws IOException, MqttException {
 		socket.setSoTimeout(this.handshakeTimeoutSecs * 1000);
 		
 		// SNI support.  Should be automatic under some circumstances - not all, apparently
-		SSLParameters sslParameters = new SSLParameters();
-		List<SNIServerName> sniHostNames = new ArrayList<SNIServerName>(1);
-		sniHostNames.add(new SNIHostName(host));
-		sslParameters.setServerNames(sniHostNames);
-		((SSLSocket)socket).setSSLParameters(sslParameters);
+		try {
+			SSLParameters sslParameters = new SSLParameters();
+			List<SNIServerName> sniHostNames = new ArrayList<SNIServerName>(1);
+			sniHostNames.add(new SNIHostName(host));
+			sslParameters.setServerNames(sniHostNames);
+			((SSLSocket)socket).setSSLParameters(sslParameters);
+		} catch(NoClassDefFoundError e) {
+			// Android < 7.0
+		}
 
 		// If default Hostname verification is enabled, use the same method that is used with HTTPS
 		if(this.httpsHostnameVerificationEnabled) {
-			SSLParameters sslParams = new SSLParameters();
-			sslParams.setEndpointIdentificationAlgorithm("HTTPS");
-			((SSLSocket) socket).setSSLParameters(sslParams);
+			try {
+				SSLParameters sslParams = new SSLParameters();
+				sslParams.setEndpointIdentificationAlgorithm("HTTPS");
+				((SSLSocket) socket).setSSLParameters(sslParams);
+			} catch(NoSuchMethodError e) {
+				// Android < 7.0
+			}
```

Comment:
Previously, this block said:
> If hostname verification is enabled, then perform host name verification as part of the handshake.

 With this change, the code now says:
> If hostname verification is enabled, then perform host name verification if we can. But if we can't, don't tell anyone, and assume the hostname was verified.

This is problematic, since you're breaking a security contract with the user. They think the hostname is verified, but it isn't.

Android documentation suggests a mechanism to do hostname verification < 24 here:
https://developer.android.com/training/articles/security-ssl#java

A better implementation here might look like:

```
if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
    // Verify hostname via SSLParamters.
    SSLParameters sslParams = new SSLParameters();				 
    sslParams.setEndpointIdentificationAlgorithm("HTTPS");
    (SSLSocket) socket).setSSLParameters(sslParams);
}
...
if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
    verifyHostname(hostname);
}
...
private void verifyHostname(String hostname) throws SSLHandshakeException {
    boolean hostnameVerified =
        HttpsURLConnection.getDefaultHostnameVerifier().verify("mail.google.com", socket.getSession());
    if (!hostnameVerified) {
        throw new SSLHandshakeException("Expected " + hostname + ", found " + s.getPeerPrincipal());
    }
}
```

As a work-around, a consuming app could arrange for `this.httpsHostnameVerificationEnabled` to be `false`, when running on < API 24.

Reply:
Good point, but `Build.VERSION.SDK_INT` and other `Build` constants are from Android SDK and it's not part of the Paho dependencies. AFAIK you can't add it as a simple depencency to Maven and I don't think you want to add the whole Android SDK to Paho just for this.

Maybe we may replace the whole thing with the `HttpsURLConnection` variant which should work everywhere.

But I'm afraid this won't work with SNI anyway, as is mentioned in the Android documentation you linked to.